## Kütüphanelerin import Edilmesi

In [7]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [8]:
%matplotlib notebook

## Sklearn Veri Seti Oluşturma

In [9]:
from sklearn.datasets import make_regression
x, y = make_regression(n_samples=100, n_features=1, noise=1, n_targets=1,random_state=42)

x = np.array(x)
y = np.array(y).reshape(-1, 1) 

## Stochastic Gradient Descent Fonksiyonu

In [10]:
def stochastic_gradient_descent(x, y, theta0, theta1, learning_rate, iterations, threshold):
    #stochastic gd nin farkı her iterasyonda veri setimizin sadece bir örneğini kullanarak parametreleri güncelleyeceğiz
    m = len(x)
    theta_history = []

    for i in range(iterations):
        for j in range(m):
            #sadece veri setimizin j örneğini hesaplayarak parametreleri hesapla
            h = theta0 + theta1 * x[j]
            error = h - y[j]
        
            theta0 = theta0 - learning_rate * error
            theta1 = theta1 - learning_rate * error * x[j]

        #hatanın thereshold dan küçük olup olmadığının kontrolü
        error_sum = np.sum((theta0 + theta1 * x - y) ** 2)
        if error_sum < threshold:
            print(f"Threshold ({threshold}) değeri aşıldı. İterasyon: {i + 1}")
            break
        
        theta_history.append((theta0, theta1))

    return theta_history


## Karşılaştırma için Gradient Descent Algoritması

In [11]:
def gradient_descent(x, y, theta0, theta1, learning_rate, iterations, threshold):
    m = len(x)
    theta_history = []
    for i in range(iterations):
        # Hesapla
        h = theta0 + theta1 * x
        error = h - y
        
        # Gradient descent adımları
        theta0 = theta0 - (learning_rate/m) * sum(error)
        theta1 = theta1 - (learning_rate/m) * sum(error * x)
        
        error_sum = np.sum(error ** 2)  # Hata karesinin toplamını hesapla
        if error_sum < threshold:
            print(f"Threshold ({threshold}) değeri aşıldı. İterasyon: {i + 1}")
            break
        
        #print(f"Iteration {i + 1}: theta0 = {theta0}, theta1 = {theta1}")
        
        theta_history.append((theta0, theta1))
    return theta_history

## Parametre Değerleri

In [12]:
initial_theta0 = 1
initial_theta1 = 1

learning_rate = 0.01
iterations = 1000
threshold = 100

In [13]:
# Görselleştirme amacıyla bütün iterasyon aşamalarındaki tetha değerleri liste olarak tutuluyor
theta_history = stochastic_gradient_descent(x, y, initial_theta0, initial_theta1, learning_rate, iterations, threshold)

final_theta0 = theta_history[(len(theta_history)-1)][0]
final_theta1 = theta_history[(len(theta_history)-1)][1]

print(f"Final theta0: {final_theta0}")
print(f"Final theta1: {final_theta1}")

Threshold (100) değeri aşıldı. İterasyon: 6
Final theta0: [-0.10644024]
Final theta1: [41.34659587]


In [14]:
theta_history = gradient_descent(x, y, initial_theta0, initial_theta1, learning_rate, iterations, threshold)

final_theta0 = theta_history[(len(theta_history)-1)][0]
final_theta1 = theta_history[(len(theta_history)-1)][1]

print(f"Final theta0: {final_theta0}")
print(f"Final theta1: {final_theta1}")

Threshold (100) değeri aşıldı. İterasyon: 545
Final theta0: [-0.05236613]
Final theta1: [41.51208528]


## Rapor
Bu rapor Stochastic Gradient Descent (SGD)  Gradient Descent (GD) algoritmalarını kullanarak gerçekleştirilen regresyon modelinin eğitimi üzerine odaklanmaktadır. İki algoritma da bir eğitim eşiği (threshold) belirlenene kadar çalıştırıldı.

1. Stochastic Gradient Descent (SGD):

Threshold (Eşik) Değeri: 100
İterasyon Sayısı: 6
Final Theta0: -0.10644024
Final Theta1: 41.34659587

2. Batch Gradient Descent (BGD):

Threshold (Eşik) Değeri: 100
İterasyon Sayısı: 545
Final Theta0: -0.05236613
Final Theta1: 41.51208528

Stochastic Gradient Descent algoritması, belirlenen eşik değerine daha hızlı ulaşarak %99'a yakın bir iterasyon sayısında düşüş sağlamıştır.Ancak bu hız daha rastgele güncellemelerin neden olduğu dalgalanmalarla birlikte gelir ve bu durum modelin hız ve kararlılığını etkileyebilir. 